man/smsAnal.1
.TH smsAnal 1 "2008 Feb 22" GNU
.SH NAME
smsAnal - SMS Analysis command line application.
.SH SYNOPSIS
.B smsAnal
[-\fIoption\fP ...]
.I inputSoundFile outputSmsFile
.SH DESCRIPTION
\fISMS\fP is a set of techniques and software implementations for the
analysis, transformation and synthesis of musical sounds based on a
sinusoidal plus residual model. These techniques can be used for
synthesis, processing and coding applications, while some of the
intermediate results might also be applied to other music related
problems, such as sound source separation, musical acoustics, music
perception, or performance analysis. The basic model and
implementation was developed in the PhD thesis by X. Serra in 1989 and
since then many extensions have been proposed at MTG-UPF and by other
researchers.

\fIsmsAnal\fP will do SMS analysis from the command line on a variety
of soundfile formats (using the libsndfile library, developed by Erik
de Castro Lopo). The output is a .sms; to see how to modify and
synthisize the analysis, see the README included with the SMS package
or visit the SMS homepage at:
\fIhttp://www.iua.upf.es/~sms/\fP
.SH OPTIONS
.B Description of parameters
.TP 8
.BI -v " verbose mode"
Print out lots of information about parameters and what is going on.
.TP 8
.BI -d " debugMode"
.B (default 0) [1,2,3,4,5,6,7,8,9,10,11,12]
0 no debug, 1 debug initialitzation functions, 2 debug peak detection function, 3 debug harmonic detection function, 4 debug peak continuation function, 5 debug clean trajectories function, 6 debug sine synthesis function, 7 debug stochastic analysis function, 8 debug stochastic synthesis function, 9 debug top level analysis function, 10 debug everything, 11 write residual into a file (residual.snd), 12 write original, synthesis and residual to a text file (debug.txt).
.TP 8
.BI -f " format"  
.B (default 0) [0,1,2,3] 
format of the representation: 0 harmonic, 1 inharmonic, 2 harmonic with phase, 3 inharmonic with phase.
.TP 8
.BI -q " soundType"
.B (default 0) [0,1]
type of sound to be analyzed. 0: sound phrase, 1: single note. Useful for single stable notes. When this is set to 1 the default fundamental (-u) is used as the reference fundamental, there is practically no pitch detection.
.TP 8
.BI -x " analysisDirection"
.B (default 0) [0,1]
direction of the analysis. 0: direct, 1: reverse. Reverse is very useful for percussive sounds or sounds with a noisy attack.
.TP 8
.B STFT parameters
.TP 8
.BI -s " windowSize"
.B (default 3.5) [>=1]
number of periods of fundamental frequency to use in the analysis window. The actual window size in seconds will be this value divided by the fundamental frequency found at every given moment.
.TP 8
.BI -i " windowType"
.B (default 1) [0,1,2,3,4]
type of analysis window to use. 0: Hamming, 1: Blackman-Harris 62 dB, 2: Blackman-Harris 70 dB, 3: Blackman-Harris 74 dB, 4: Blackman-Harris 92 dB.
.TP 8
.BI -r " frameRate"
.B (default 300) [50 <-> 600]
number of analysis windows per second (Hz). This will determine the hop size of the analysis window. If given as a negative number this value will be the overlap factor, and the frame rate will be calculated from that.
.TP 8
.B Peak detection parameters
.TP 8
.BI -j " highestFreq"
.B (default 12000) [20 <-> 22500]
highest frequency in Hz of the peaks to be detected. Therefore no partials higher than this frequency will be detected. It will never be higher than half the sampling-rate.
.TP 8
.BI -k " minPeakMag"
.B (default 0) [0. <-> 100.]
minimum magnitude in decibels of a peak. Peak softer than this dB value will not have any chance to be considered part of the deterministic component, that is, of the partials. This value should not be smaller than 0 since 0 is the noise threshold used in the analysis.
.TP 8
.B Harmonic detection parameters
.TP 8
.BI -y " refHarmonic"
.B (default 1) [0,1,2,3 ...]
number of the harmonic used for reference, 1 is the fundamental. There are some sounds, like many piano sounds, that have a very soft fundamental. In these cases it is helpful to find the fundamental frequency by looking for a harmonic other than the actual fundamental.
.TP 8
.BI -m " minRefHarmMag"
.B (default 30) [5 <-> 60]
minimum magnitude in dB of the harmonic used for reference in the harmonic detection process. 
.TP 8
.BI -z " refHarmDiffFromMax"
.B (default 30) [5 <-> 60]
maximum dB difference between the harmonic used for reference and the maximum peak.
.TP 8
.BI -u " defaultFund"
.B (default 100) [20 <-> 5000]
default fundamental frequency in Hz. This is the frequency that is used to set the actual analysis window size when no fundamental has been found. In normal situations it is convenient to give the value of the fundamental frequency of the begining of the sound so that it can start with a good guess. In the case of inharmonic sounds this value will be used to set the window size for the whole sound. When defaultFund is higher than highestFund it is set to this value and when it is lower than lowestFund it is set to it.
.TP 8
.BI -l " lowestFund"
.B (default 50) [20 <-> 5000]
In harmonic signals, this is the lowest fundamental frequency in Hz to be searched for. In the case of inharmonic sounds this value is used as the lowest frequency to track. No peak is found below this value.
.TP 8
.BI -h " highestFund"
.B (default 1000) [20 <-> 5000]
highest fundamental frequency in Hz to be searched for. Only used in harmonic sounds.
.TP 8
.B Peak continuation parameters
.TP 8
.BI -n " nGuides"
.B (default 100) [1 <-> 500]
number of guides to be used in analysis. These guides will be used to track the partials in the sound and are the ones that will be subtracted from the original sound. The number of output tracks is defined by the parameter nTracks.
.TP 8
.BI -p " nTracks"
.B (default 60) [1 <-> 500]
maximum number of tracks, or partials, to be found. This will be the output number of tracks. 
.TP 8
.BI -w " freqDeviation"
.B (default .45) [.1 <-> .5]
maximum deviation that is permitted from the "guide frequency" to the continuation peak of the guide. In the case of harmonic sounds the deviation in Hz is the product of this value times the fundamental frequency. In the case of inharmonic sounds the deviation in Hz is this value times the guide frequency.
.TP 8
.BI -t " peakContToGuide"
.B (default .4) [0 <-> 1]
contribution of the frequency of the previous peak of a given trajectory to the current guide frequency value. If the value is 1, it means that the previous peak will completely define the guide value, the possible current fundamental will not be used to set the guide's frequency. If the value is 0, the previous guide will not be used at all.
.TP 8
.BI -o " fundContToGuide"
.B (default .5) [0 <-> 1]
contribution of the fundamental frequency of the current frame to the current guide frequency. This is only relevant in harmonic sounds.
.TP 8
.B Trajectory cleaning paramenters
.TP 8
.TP 8
.BI -g " cleanTraj"
.B (default 1) [0,1]
whether or not to clean the deterministic data after analysis. 0 no cleaning, 1 cleaning. This cleaning process gets rid of short trajectories that may not be part of a stable partial of the sound and also fill gaps in stable partials.
.TP 8
.BI -a " minTrajLength"
.B (default .1) [0 <-> 10]
minimum length of the trajectories in seconds. Trajectories shorter than this value will be deleted if the cleanTraj flag (-g) has been set.
.TP 8
.BI -b " maxSleepingTime"
.B (default .1) [0 <-> .5]
maximum sleeping time in seconds for a given trajectory. Time shorter than this value will be considered gaps in the trajectory and if the cleanTraj flag (-g) has been set, this gaps will be filled by interpolating the boundaries.
.TP 8
.B Stochastic analysis parameters
.TP 8
.BI -stochastic " stochasticType"
.B (default 1) [0,1,2]
type for the stochastic representation: (0) none,  (1) line segments on magnitude spectrum, or (2) STFT real/imaginary pairs.
The first time the analysis is done it is useful to set this to 3 (this will let you check if the analysis was well done and the computation time will be much shorter).  
(2) is not currently implemented and defaults to (1)
.TP 8
.BI -c " nStocCoeff"
.B (default 128) [min 4]
When the stochastic type is set to 1(line segments on magnitude spectrum), this number corresponds to the number of inflexion points (will be interpolated to the necessary number of bins for the stochastic IFFT). The actual number of coefficients is limited to 1/2 the size of the FFT used  to create the spectrum ( samperate / framerate, roundup up to a power of 2).
.TP 8
.B spectral enveloping parameters
.TP 8
.BI --se " spectral envelope type"
.B (default 0) [0,1,2] 
0: off, 1: cepstrum coefficients, 2: frequency bins (converted from cepstrum).
.TP 8
.BI --co " discrete cepstrum order"
.B (default 25) 
The order of cepstrum, order+1 cepstrum coefficients will be produced.
.TP 8
.BI --la " lambda"
.B (default 0.00001) 
A regularizing coefficient to help make the cepstrum spectral envelope behave in some situations.  Increase this to reduce oscillations, if they occur.
.TP 8
.BI --an " anchor"
.B (default 0, off) 
Turn on envelope anchoring, which will duplicate add two more sinusoidal peaks to the ends of the spectrum in order to prevent certain peaks from getting cut off by the envelope.
.TP 8
.BI --mef " maximum envelope frequency"
.B (default highest freqequency in peak detection) 
Define the maximum envelope frequency.  You normally want to leave this at the default, unless you are combining with a sound that has a different maximum envelope frequency.
.SH SEE ALSO
smsSynth(1), smsClean(1), smsPrint(1), smsResample(1)


doc/changes.txt
=-=-=-=-=-=-=- libsms change log -=-=-=-=-=-=

=== June 23, 2009 === version 1.1 beta =====================================================

New features:
	- Can now transform sounds using their spectral envelopes. By setting the appropriate
	flags in an SMS_ModifyParams struct and calling sms_modify, it is possible to shape the
	partials of one sound using the spectral envelope of another (or a linear interpolation
	anywhere between their two envelopes), and to transpose sounds while maintaining the
	original spectral envelope. In fact the envelopes do not have to be related to any
	analysed sounds, they could be generated by any means and then applied to the detected
	partials of a given sound, so purely synthetic envelopes could be used as well.
	- The full SMS analysis/synthesis process is now callable from Python, and a Python
	module called SMS.py has been created that will analyze or synthesize a file with a single
	function call. See the demo script python/AnalysisSynthesis.py for an example of using it.
        - new misc. functions: sms_rms, sms_arrayScalarTempered

API changes:
	- All transformations (including transpostion and stochastic component scaling) should
	now be performed using the sms_modify function. The required transformations are specified
	by using the SMS_ModifyParams structure. Examples of using this (from Python) can be found
	in the file python/modify_examples.py.
        - dB conversion is slightly different.  It is normalized to above 0, using a threshhold (default 0.00001)
          to avoid NaN's.  You can set this with sms_setMagThresh
        - all magnitudes are now stored in linear
        - windows are now left un-normalized by default. Call sms_scaleWindow() to normalize it.
Bug fixes:
	- The stochastic gain calculated during residual analysis was being applied twice during
	resynthesis, fixed this.
	- Also changed residual and spectrum energy calculation in sms_residual so that they use 
	the sum of the square of signal samples rather than just the sum of the samples.
	- Fixed a call to memset is sms_synthesize that was writing more data than intended (writing 
	more than the given array size)
	- made some changes to python/setup.py so that the python module builds on OS X (tested
	on Leopard).

Other:
	- Added some unit tests (in Python) to test/Tests.py
        - smsToYaml is gone. smsPrint now naturally prints in yaml format, and you can save it to file.

=== April 9, 2009 === version 1.05e, experimental release =====================================
This is a release heading in the direction of better modularity, especially in the analysis routines.  The ultimate
goal is to be able to interchange each step of the analysis process (such as f0 estiumation, transient extraction,
etc) with custom code, either in C or python.

Many functions received argument changes so they could be easily wrapped in python using SWG.  I tried to
give every array pointer an argument before it that indicates the size of the array.  In some functions, this seemed
too redundant (such as sms_spectrum for example), so they are custom wrapped in the swig interface file.

In doing this, the release is made incompatible with the last because of many name and argument changes. This will
probably continue to happen through the summer.  If projects break or people think things were better the way before,
please email me.

Notable changes:
        - some files have been moved around in an attempt to organize function types (transforms.c. filters.c, tables.c, etc)
        - some more general functions are in sms.c, such as RMS array calculation and frequency conversion (well-tempored)
        - started replacing float with sfloat so that, one day, the library can be optionally compiled in double precision
        - added support for reading of multi-channel audio files - can select what channel you want to analyze 
        - smsAnal now depends much more on library functions (like sms_initAnalParams)
        - tried to fix documentation errors mentioned to me
        - removed some outdated tools (smsToLisp, smsSynthDet - now in smsSynth)
        - moved around synthesis buffers so that they are properly re-allocated if the sizeHop changes
        - using Mersainne Twister random number generator.
        - sms_analyzeFrame() is supposed to be a function for computing sinusoidal analysis without a circular buffer.
          Its not quite at that point yet.
        - SMS_PeakParams structure has parameters just for sinusoidal analysis (it used to need all of SMS_AnalParams)
        - python test scripts have been added for many things (all names test_*.py)

=== Nov. 15, 2008 === version 1.0, first release ============================================

This release represents a revival of the original code from Xavier Serra, which dates back to the early 1990's.
Care has been taken to make sure the functionality of the algorithms important to SMS have not changed,
while still updating function calls to work with modern day platforms.  Future releases will focus more on incorporating
improvements in the SMS model that have since been discovered. Some changes to the original code include:
        - using libsndfile for sound file handling
        - using OOURA fft routines for real-DFT functions (fftw3 as well, but not fully working)
        - optionally can compile to use FFTW3 fft libary, but it is still in a testing stage and fails in pd externals
        - appended the namespaces sms_* and SMS_ to all globally seen functions, structures and defines/enumerations
        - uses scons instead of make, as it is very powerful, being based in python
        - python SWIG wrappers
        - synthesis parameters "stocgain" and "ftranspose" are added to change the current synthesis, but not the data itself
        - smsToYaml - converts a binary *.sms file to YAML text format, which is very easy to read

tools/smsAnal.c
/* 
 * Copyright (c) 2008 MUSIC TECHNOLOGY GROUP (MTG)
 *                         UNIVERSITAT POMPEU FABRA 
 * 
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 */
#include "sms.h"
#include <popt.h>

const char *help_header_text =
        "\n\n"
        "Usage: smsAnal [options]  <inputSoundFile> <outputSmsFile>\n"
        "\n"
        "analyzes a sound file a stores it in a binary SMS file. All parameters are numbers. "
        "See the man page for details.\n\n"
        "Options:\n";

int main (int argc, const char *argv[])
{
	FILE *pOutputSmsFile; 
	SMS_Data smsData;
	SMS_Header smsHeader;

	float pSoundData[SMS_MAX_WINDOW];
	SMS_SndHeader soundHeader;

	char *pChInputSoundFile = NULL, *pChOutputSmsFile = NULL;
        int verbose = 0;
	//int iHopSize;
	int iDoAnalysis = 1;
        int iFrame = 0;
	long iStatus = 0, iSample = 0, sizeNewData = 0;

	int optc;   /* switch */
	poptContext pc;

	SMS_AnalParams analParams;
        sms_initAnalParams(&analParams); 	/* initialize arguments to defaults*/

        struct poptOption options[] =
                {
                        {"verbose", 'v', POPT_ARG_NONE, &verbose, 0, 
                         "verbose mode", 0},
                        {"debug", 'd', POPT_ARG_INT, &analParams.iDebugMode, 0, 
                         "debug mode (0)", "int"},
                        {"format", 'f', POPT_ARG_INT, &analParams.iFormat, 0, 
                         "analysis format (0, harmonic)", "int"},
                        {"sound-type", 'q', POPT_ARG_INT, &analParams.iSoundType, 0, 
                         "sound type (0, phrase)", "int"},
                        {"direction", 'x', POPT_ARG_INT, &analParams.iAnalysisDirection, 0, 
                         "analysis direction (0, forward)", "int"},
                        /* STFT Parameters: */
                        {"window-size", 's', POPT_ARG_FLOAT, &analParams.fSizeWindow, 0, 
                         "size of the window in f0 periods (3.5)", "float"},
                        {"window-type", 'i', POPT_ARG_INT, &analParams.iWindowType, 0, 
                         "window type (1, blackman harris 70 dB)", "int"},
                        {"frame-rate", 'r', POPT_ARG_INT, &analParams.iFrameRate, 0, 
                         "frame rate in hertz (300)", "int"},
                        /* Peak Detection Parameters */
                        {"highest-freq", 'j', POPT_ARG_FLOAT, &analParams.fHighestFreq, 0, 
                         "highest frequency to look for peaks (12000hz)", "float"},
                        {"min-peak-mag", 'k', POPT_ARG_FLOAT, &analParams.fMinPeakMag, 0, 
                         "minimum peak magnitude (0 normalized dB, which corresponds to -100dB)", "float"}, /*\todo check this doc*/
                        /* Harmonic Detection Parameters */
                        {"ref-harmonic", 'y', POPT_ARG_INT, &analParams.iRefHarmonic, 0, 
                         "reference harmonic number in series (1)", "int"},
                        {"min-ref-harm-mag", 'm', POPT_ARG_FLOAT, &analParams.fMinPeakMag, 0, 
                         "minimum reference harmonic magnitude (30 normalized dB)", "float"}, /*\todo check this doc*/
                        {"min-ref-harm-mag", 'z', POPT_ARG_FLOAT, &analParams.fRefHarmMagDiffFromMax, 0, 
                         " maximum dB difference between the harmonic used for reference and the maximum peak (default 30)", "float"}, /*\todo check this doc*/
                        {"default-fund", 'u', POPT_ARG_FLOAT, &analParams.fDefaultFundamental, 0, 
                         "default fundamental frequency (hz), used to set initial window size, or window size for entire sound if inharmonic (default 100)", "float"}, 
                        {"lowest-fund", 'l', POPT_ARG_FLOAT, &analParams.fDefaultFundamental, 0, 
                         "lowest fundamental frequency(hz), or frequency in inharmonic analysis, to search for (default 50)", "float"}, 
                        {"highest-fund", 'h', POPT_ARG_FLOAT, &analParams.fDefaultFundamental, 0, 
                         "highest fundamental frequency to search for, has no effect on inharmonic analysis (default 1000)", "float"}, 
                        /* Peak Continuation parameters */
                        {"guides", 'n', POPT_ARG_INT, &analParams.nGuides, 0, 
                         "number of guides to use in partial tracking (default 100)", "int"},
                        {"tracks", 'p', POPT_ARG_INT, &analParams.nTracks, 0, 
                         "number of output partial tracks (default 60)", "int"},
                        {"freq-deviation", 'w', POPT_ARG_FLOAT, &analParams.fFreqDeviation, 0, 
                         "maximum permitted frequency deviation from guide frequency (default .45)", "float"}, 
                        {"peak-cont-guide", 't', POPT_ARG_FLOAT, &analParams.fPeakContToGuide, 0, 
                         "contribution of the frequency of the previous peak of a given trajectory to the current guide frequency value (default .4).", "float"}, 
                        {"fund-cont-guide", 'o', POPT_ARG_FLOAT, &analParams.fFundContToGuide, 0, 
                         "contribution of the fundamental frequency of the previous peak of a given trajectory to the current guide frequency value (default .5).", "float"}, 
                        /* Track Cleaning parameters:\n" */
                        {"clean-track", 'g', POPT_ARG_INT, &analParams.iCleanTracks, 0, 
                         "turn on/off track cleaning (default is on, 1)", "int"}, 
                        {"min-track-length", 'a', POPT_ARG_FLOAT, &analParams.iMinTrackLength, 0, 
                         "minimum track length in seconds (0.1)", "float"}, 
                        {"max-sleeping-time", 'b', POPT_ARG_FLOAT, &analParams.iMaxSleepingTime, 0, 
                         "maximum time a frame can sleep in seconds (0.1)", "float"},  /* this doc is horrible */
                        /* Stochastic Analysis parameters */
                        {"stochastic", 'e', POPT_ARG_INT, &analParams.iStochasticType, 0, 
                         "turn on/off stochastic analysis (default is on, 1)", "int"}, 
                        {"stoch-coeff", 'c', POPT_ARG_INT, &analParams.nStochasticCoeff, 0, 
                         "number of stochastic coefficients in approximation (default 128)", "int"}, 
                        /* spectral enveloping parameters */
                        {"se",0, POPT_ARG_INT, &analParams.specEnvParams.iType, 0, 
                         "spectral enveloping type (0, off)", "int"},
                        {"co", 0, POPT_ARG_INT, &analParams.specEnvParams.iOrder, 0, 
                         "discrete cepstrum order (25)", "int"},
                        {"la", 0, POPT_ARG_FLOAT, &analParams.specEnvParams.fLambda, 0, 
                         "lambda, regularizing coefficient (0.00001)", "float"}, 
                        {"an", 0, POPT_ARG_NONE, &analParams.specEnvParams.iAnchor, 0, 
                         "turn on anchoring of spectral envelope endpoints", 0}, 
                        {"mef", 0, POPT_ARG_INT, &analParams.specEnvParams.iMaxFreq, 0, 
                         "maximum envelope frequency (default is highest-freq", "int"}, 
                        POPT_AUTOHELP
                        POPT_TABLEEND
                };


        pc = poptGetContext("smsAnal", argc, argv, options, 0);
        poptSetOtherOptionHelp(pc, help_header_text);

        while ((optc = poptGetNextOpt(pc)) > 0) {
                switch (optc) {
                        /* specific arguments are handled here */
                case 'v':
                        verbose = 1;
                default:
                        ;
                }
        }
        if (optc < -1) 
        {
                /* an error occurred during option processing */
                printf("%s: %s\n",
                       poptBadOption(pc, POPT_BADOPTION_NOALIAS),
                       poptStrerror(optc));
                return 1;
        }
        if (argc < 3)
        {
                poptPrintUsage(pc,stderr,0);
                return 1;
        }
	pChInputSoundFile = (char *) poptGetArg(pc);
	pChOutputSmsFile = (char *) poptGetArg(pc);
        /* parsing done */


	/* open input sound */
	if (sms_openSF(pChInputSoundFile, &soundHeader))
	{
                printf("error in sms_openSF: %s \n", sms_errorString());
                exit(EXIT_FAILURE);
	}	    

        /* initialize everything */
        sms_init();
        /* TODO NExt: go from here through all the functions that need to look at specEnvParams */
        sms_initAnalysis (&analParams, &soundHeader);
        
	sms_fillHeader (&smsHeader, &analParams, "smsAnal");
	sms_writeHeader (pChOutputSmsFile, &smsHeader, &pOutputSmsFile);

        /* allocate output SMS record */
	sms_allocFrameH (&smsHeader, &smsData);

	/* perform analysis */

	if (analParams.iAnalysisDirection == SMS_DIR_REV)
		iSample = soundHeader.nSamples;

	if (analParams.iDebugMode == SMS_DBG_SYNC)
		sms_createDebugFile (&analParams);
	if (analParams.iDebugMode == SMS_DBG_RESIDUAL)
		sms_createResSF (analParams.iSamplingRate);

        if(verbose)
        {
                printf("\n===sound file info===\n");
                printf("samples: %d, samplerate: %d, seconds: %f \n", soundHeader.nSamples, soundHeader.iSamplingRate, 
                       soundHeader.nSamples / (float)soundHeader.iSamplingRate);
                printf("number of channels: %d, read channel: %d \n", soundHeader.channelCount, soundHeader.iReadChannel); 
                printf("\n===analysis parameters===\n");
		printf("sizeHop: %d, nFrames: %d \n", analParams.sizeHop, analParams.nFrames);
                /* \todo: print analysis window type (by name) here */
                if(analParams.specEnvParams.iType != SMS_ENV_NONE)
                {
                        printf("\n===spectral envelope parameters===\n");
                        if(analParams.specEnvParams.iType == SMS_ENV_CEP)
                                printf("type: cepstral coefficients, ");
                        else if(analParams.specEnvParams.iType == SMS_ENV_FBINS)
                                printf("type: frequency bins, ");
                        else
                                printf("warning: unknown spectral envelope type! \n\n ");
                        printf("order: %d, lambda: %f, max frequency: %d \n", analParams.specEnvParams.iOrder,
                               analParams.specEnvParams.fLambda, analParams.specEnvParams.iMaxFreq);
                }
		printf("\n===header info string===\n %s", smsHeader.pChTextCharacters);
                printf("\n\ndoing analysis now:\n");
        }

	while(iDoAnalysis > 0)
	{
		if (analParams.iAnalysisDirection == SMS_DIR_REV)
		{
			if ((iSample - analParams.sizeNextRead) >= 0)
				sizeNewData = analParams.sizeNextRead;
			else
				sizeNewData = iSample;
                        iSample -= sizeNewData;
		}
		else
		{
			iSample += sizeNewData;
			if((iSample + analParams.sizeNextRead) < soundHeader.nSamples)
				sizeNewData = analParams.sizeNextRead;
			else
				sizeNewData = soundHeader.nSamples - iSample;
		}
		/* get one frame of sound */
		if (sms_getSound (&soundHeader, sizeNewData, pSoundData, iSample))
		{
			printf("error: could not read sound frame %d\n", iFrame);
                        printf("error message in sms_getSound: %s \n", sms_errorString());
			break;
		}
		/* perform analysis of one frame of sound */
		iStatus = sms_analyze (sizeNewData, pSoundData, &smsData,
		                       &analParams);

		/* if there is an output SMS record, write it */
		if (iStatus == 1)
		{
			sms_writeFrame (pOutputSmsFile, &smsHeader, &smsData);
                        if(sms_errorCheck())
                        {
                                printf("error: could not write sms frame %d:\n", iFrame);
                                printf("error message in sms_writeFrame: %s \n", sms_errorString());
                                break;
                        }
			if(verbose)
                        {
/*                                 if (iFrame % 10 == 0) */
/*                                         printf ("frame: %d, %.2f \n", iFrame, iFrame / (float) smsHeader.iFrameRate); */
                                if (iFrame % 10 == 0)
                                        printf ("%.2f ", iFrame / (float) smsHeader.iFrameRate);

                        }
			iFrame++;
		}
		else if (iStatus == -1) /* done */
		{
			iDoAnalysis = 0;
			smsHeader.nFrames = iFrame;
		}
	}


        smsHeader.fResidualPerc = analParams.fResidualAccumPerc / iFrame;
        if(verbose)
        {
                printf("\n");
                printf("residual percentage: %f \n", smsHeader.fResidualPerc);
        }                
        if(smsHeader.nFrames != analParams.nFrames && verbose)
                printf("warning: wrong number of analyzed frames: analParams: %d, smsHeader: %d \n", 
                       analParams.nFrames, smsHeader.nFrames); 
	/* write an close output files */
	sms_writeFile (pOutputSmsFile, &smsHeader);
	if (analParams.iDebugMode == SMS_DBG_RESIDUAL)
		sms_writeResSF ();
	if (analParams.iDebugMode == SMS_DBG_SYNC)
		sms_writeDebugFile ();

        printf("wrote %d analysis frames to %s\n", iFrame, pChOutputSmsFile);

        /* cleanup */
        sms_freeFrame(&smsData);
        sms_freeAnalysis(&analParams);
        sms_free();
	return 0;	
}

